src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java

```diff
@@ -57,57 +55,92 @@ public static void setupLocks() {
-        for (Ingredient ingredient : craftTweakerLocks.keySet()){
-            ItemStack[] ingredientMatchingStacks = ingredient.getMatchingStacks();
-            for (int i = 0; i < ingredientMatchingStacks.length; i++){
-                compareItem(ingredientMatchingStacks[i], stack);
+
+        ItemInfo cleanStack = new ItemInfo(stack.getItem(), stack.getMetadata());
+        //There is no NBT information so matching is not needed. OR no specific NBT locks for this item stack so the NBT tags can be ignored
+        if (!stack.hasTagCompound() || !nbtLockInfo.containsKey(cleanStack))
+            return locks.getOrDefault(cleanStack, EMPTY_LOCK);
+
+        NBTTagCompound tag = stack.getTagCompound();
+        Set<NBTTagCompound> nbtLookup = nbtLockInfo.get(cleanStack);
+        ItemInfo bestStack = cleanStack; //Default to the clean variant of the stack
+        for (NBTTagCompound nbt : nbtLookup) {
+            if (tag.equals(nbt)) {
+                bestStack = new ItemInfo(stack.getItem(), stack.getMetadata(), nbt);
+                break;
             }
         }
-        return EMPTY_LOCK;
+        return locks.getOrDefault(bestStack, EMPTY_LOCK);
     }
 
-    private static boolean compareItem(ItemStack itemStack, ItemStack stack) {
-        if (itemStack.hasTagCompound() != stack.hasTagCompound()){
-            return false;
-        }
-        if (itemStack.hasTagCompound() && stack.hasTagCompound()) {
-            if (itemStack.getTagCompound().getKeySet().equals(stack.getTagCompound().getKeySet())){
-                for (String s : itemStack.getTagCompound().getKeySet()){
-                    if (!itemStack.getTagCompound().getTag(s).equals(stack.getTagCompound().getTag(s))){
-                        return false;
-                    }
-                }
-            }
-        }
-        if (itemStack.hasTagCompound() && !stack.hasTagCompound()){
-            return stack.getItem() == itemStack.getItem() && (itemStack.getMetadata() == 32767 || stack.getMetadata() == itemStack.getMetadata());
+    private static class ItemInfo {
+        private int metadata;
+        private Item item;
+        private NBTTagCompound tag;
+
+        private ItemInfo(Item item, int metadata) {
+            this(item, metadata, null);
         }
-        if (!itemStack.hasTagCompound() && stack.hasTagCompound()){
-            return stack.getItem() == itemStack.getItem() && (itemStack.getMetadata() == 32767 || stack.getMetadata() == itemStack.getMetadata());
+
+        private ItemInfo(Item item, int metadata, NBTTagCompound tag) {
+            this.item = item;
+            this.metadata = metadata;
+            this.tag = tag;
         }
-        if (!itemStack.hasTagCompound() && !stack.hasTagCompound()){
-            return stack.getItem() == itemStack.getItem() && (itemStack.getMetadata() == 32767 || stack.getMetadata() == itemStack.getMetadata());
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == this)
```

Comment:
Always do ```if (o == this) {```

Reply:
I will implement that change, however I am curious as to why you say to always do it. Is there a specific reason or is it just a syntax that you prefer having brackets for single line if statements or just this if statement in general for equals methods.

Reply:
General coding style preferences for all things.